File: bin/database_build.py
#!/usr/bin/python3
from pickle import TRUE
import json
import pymysql
import sys
import os
from dotenv import load_dotenv
load_dotenv()  # Load database name from the .env file

sys.path.append('/opt/software-discovery-tool/src/config')
import supported_distros
SUPPORTED_DISTROS = supported_distros.SUPPORTED_DISTROS
SDT_BASE = '/opt/software-discovery-tool'
DATA_FILE_LOCATION = '%s/distro_data/data_files' % SDT_BASE

HOST = os.environ.get('DB_HOST')
USER = ''
PASSWORD = ''
DB_NAME = os.environ.get('DB_NAME')

def connectdb(username,password,database):
    conn = pymysql.connect(host=HOST,user=username,password=password)
    cur = conn.cursor()
    query = f"CREATE OR REPLACE DATABASE {database}"
    cur.execute(query)
    conn.close()
    print("DB INITIATILIZED SUCCESSFULLY")

def db_init():
    username = USER
    password = PASSWORD
    table_name = ""

    if password == "":
        password = input("Enter password for connecting to MariaDB server : ")
    dbName = DB_NAME

    if table_name == "" or table_name == "all" or table_name == "All":
        connectdb(username,password,dbName)
        initall(dbName,username,password)
    else:
        create_one(dbName,username,password,table_name)

def jsontosql(db,table,file,os,user,password):
    filepath = f'{DATA_FILE_LOCATION}/{file}.json'
    jsonFile = open(file=filepath)
    data = json.load(jsonFile)
    final_data = []
    for item in data :
        if item : 
            row = dict(item, osName = os)
            final_data.append(row)
    keyset = data[0].keys()
    conn = pymysql.connect(host=HOST,user=user,password=password,database=db,autocommit=True)
    curr = conn.cursor()
    try:
        data.remove({})
    except ValueError:
        pass
    if len(keyset)==2:
        query = f"INSERT INTO {table} (packageName, version, osName) VALUES (%(packageName)s, %(version)s, %(osName)s)"
    else:
        query = f"INSERT INTO {table} (packageName, version, description, osName) VALUES (%(packageName)s, %(version)s, %(description)s, %(osName)s)"
    if len(final_data) == 0 : 
        print(f"{table} : No Entries found")
    else :
        curr.executemany(query,final_data)
        print(f"{table} : Entries filled")
    conn.close()

def createTable(db,tblname,username,password):
    conn = pymysql.connect(host=HOST,user=username,password=password,database=db)
    curr = conn.cursor()
    query = f"CREATE OR REPLACE TABLE {tblname} ("\
			"pkgId INT NOT NULL AUTO_INCREMENT, "\
			"packageName VARCHAR(100) NOT NULL,"\
			"version VARCHAR(500) NOT NULL,"\
			"description VARCHAR(500), "\
			"osName VARCHAR(100) NOT NULL, "\
			"PRIMARY KEY (pkgId)"\
			")"
    curr.execute(query)
    conn.close()
    print(f"{tblname} formed successfully")

def create_one(db,username,password,table_name): 
    conn = pymysql.connect(host=HOST,user=username,password=password)
    cur = conn.cursor()
    query = f"CREATE DATABASE IF NOT EXISTS {db}"
    cur.execute(query)

    flag=True
    for os in SUPPORTED_DISTROS:
        if not SUPPORTED_DISTROS[os]:
            continue
        else:
            for distro in SUPPORTED_DISTROS[os]:
                if SUPPORTED_DISTROS[os][distro] == table_name:
                    flag=False
                    createTable(db,SUPPORTED_DISTROS[os][distro],username,password)
                    jsontosql(db,SUPPORTED_DISTROS[os][distro],SUPPORTED_DISTROS[os][distro],distro,username,password)
    conn.close()

    if flag==False:
        print(f"SUCCESSFULLY INITIALIZED {table_name} TABLE")
    else:
        print(f"NO MATCHES FOUND. TRY SOMETHING ELSE")

def initall(db,username,password): 
    count = 0
    for os_Key in SUPPORTED_DISTROS:
        if not SUPPORTED_DISTROS[os_Key]:
            continue
        else:
            for distro in SUPPORTED_DISTROS[os_Key]:
                path = f'{DATA_FILE_LOCATION}/{SUPPORTED_DISTROS[os_Key][distro]}.json'
                if os.path.exists(path):
                    createTable(db,SUPPORTED_DISTROS[os_Key][distro],username,password)
                    jsontosql(db,SUPPORTED_DISTROS[os_Key][distro],SUPPORTED_DISTROS[os_Key][distro],distro,username,password)
                    count = count+1 
                else:
                    print(f"{SUPPORTED_DISTROS[os_Key][distro]} FILE DOESN'T EXIST")
    print(f"SUCCESSFULLY INITIALIZED {count} TABLES")

if __name__ == "__main__":
    db_init()

File: bin/database_build.js
import mysql from 'mysql2/promise';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import readline from 'readline';
import dotenv from 'dotenv';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const SDT_BASE = path.join(__dirname, '..');
dotenv.config({ path: path.join(SDT_BASE, 'backend', '.env') });

const distrosPath = path.join(SDT_BASE, 'config', 'distros.json');
const SUPPORTED_DISTROS = JSON.parse(fs.readFileSync(distrosPath, 'utf8'));
const DATA_FILE_LOCATION = path.join(SDT_BASE, 'distro_data', 'data_files');

const HOST = process.env.DB_HOST || 'localhost';
const DB_NAME = process.env.DB_NAME || 'sdtDB';

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

const question = (query) => new Promise((resolve) => rl.question(query, resolve));

async function createTable(connection, tblname) {
  const queryDrop = `DROP TABLE IF EXISTS \`${tblname}\``;
  const queryCreate = `CREATE TABLE \`${tblname}\` (
    pkgId INT NOT NULL AUTO_INCREMENT,
    packageName VARCHAR(100) NOT NULL,
    version VARCHAR(500) NOT NULL,
    description VARCHAR(500),
    repo VARCHAR(500),
    osName VARCHAR(100) NOT NULL,
    PRIMARY KEY (pkgId)
  )`;
  await connection.query(queryDrop);
  await connection.query(queryCreate);
  console.log(`${tblname} formed successfully`);
}

async function jsonToSql(connection, table, file, osName) {
  const filepath = path.join(DATA_FILE_LOCATION, `${file}.json`);
  if (!fs.existsSync(filepath)) {
    console.log(`${file} FILE DOESN'T EXIST`);
    return;
  }
  const data = JSON.parse(fs.readFileSync(filepath, 'utf8'));
  const finalData = data
    .filter(item => Object.keys(item).length > 0)
    .map(item => [
      item.packageName,
      item.version,
      item.description || null,
      item.repo ? [...new Set(item.repo.split(','))].join(',') : null,
      osName
    ]);

  if (finalData.length === 0) {
    console.log(`${table} : No Entries found`);
    return;
  }

  const query = `INSERT INTO \`${table}\` (packageName, version, description, repo, osName) VALUES ?`;
  await connection.query(query, [finalData]);
  console.log(`${table} : Entries filled`);
}

async function dbInit() {
  const username = await question("Enter privileged username to create/update SQL tables: ");
  const password = await question("Enter password for privileged username: ");
  
  let connection;
  try {
    connection = await mysql.createConnection({
      host: HOST,
      user: username,
      password: password
    });

    await connection.query(`DROP DATABASE IF EXISTS \`${DB_NAME}\``);
    await connection.query(`CREATE DATABASE \`${DB_NAME}\``);
    console.log("DB INITIALIZED SUCCESSFULLY");
    await connection.changeUser({ database: DB_NAME });

    let count = 0;
    for (const osKey in SUPPORTED_DISTROS) {
      const versions = SUPPORTED_DISTROS[osKey];
      for (const distro in versions) {
        const tableName = versions[distro];
        if (tableName) {
          await createTable(connection, tableName);
          await jsonToSql(connection, tableName, tableName, distro);
          count++;
        }
      }
    }
    console.log(`SUCCESSFULLY INITIALIZED ${count} TABLES`);
  } catch (err) {
    console.error("Error during DB initialization:", err);
  } finally {
    if (connection) await connection.end();
    rl.close();
  }
}

dbInit();
